jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/AbstractHTTP2ServerConnectionFactory.java

```diff
@@ -296,22 +300,25 @@ protected ServerParser newServerParser(Connector connector, ServerParser.Listene
     }
 
     @ManagedObject("The container of HTTP/2 sessions")
-    public static class HTTP2SessionContainer implements Connection.Listener, Dumpable
+    public static class HTTP2SessionContainer implements Connection.Listener, Graceful, Dumpable
     {
-        private final Set<Session> sessions = ConcurrentHashMap.newKeySet();
+        private final Set<ISession> sessions = ConcurrentHashMap.newKeySet();
```

Comment:
As we discussed, if you protected this `sessions` collection with a `ReentrantReadWriteLock` using the read lock to modify the collection and the write lock to shut it down, you could easily guarantee that each session's `shutdown` method would only ever be called once, and be done with that `AtomicBiInteger` logic.

This would make this code a ton simpler to maintain in the long run.

Reply:
Not implementing `hashCode()` and `equals()` is by design, I really want session objects to be unique. I'll give rwlock a try.